R/step-call-pivot_wider.R

```diff
@@ -0,0 +1,225 @@
+#' # Fill in missing values
+#' fish_encounters_dt %>%
+#'   pivot_wider(names_from = station, values_from = seen, values_fill = 0)
+#'
+#' # Generate column names from multiple variables
+#' us_rent_income_dt <- lazy_dt(us_rent_income)
+#' us_rent_income_dt
+#' us_rent_income_dt %>%
+#'   pivot_wider(names_from = variable, values_from = c(estimate, moe))
+#'
+#' # When there are multiple `names_from` or `values_from`, you can use
+#' # use `names_sep` or `names_glue` to control the output variable names
+#' us_rent_income_dt %>%
+#'   pivot_wider(
+#'     names_from = variable,
+#'     names_sep = ".",
+#'     values_from = c(estimate, moe)
+#'   )
+#'
+#' # Can perform aggregation with values_fn
+#' warpbreaks_dt <- lazy_dt(as_tibble(warpbreaks[c("wool", "tension", "breaks")]))
+#' warpbreaks_dt
+#' warpbreaks_dt %>%
+#'   pivot_wider(
+#'     names_from = wool,
+#'     values_from = breaks,
+#'     values_fn = mean
+#'   )
+# exported onLoad
+pivot_wider.dtplyr_step <- function(data,
+                                    id_cols = NULL,
+                                    names_from = name,
+                                    names_prefix = "",
+                                    names_sep = "_",
+                                    names_glue = NULL,
+                                    names_sort = FALSE,
+                                    names_repair = "check_unique",
+                                    values_from = value,
+                                    values_fill = NULL,
+                                    values_fn = NULL,
+                                    ...) {
+
+  sim_data <- simulate_vars(data)
+  names_from <- names(tidyselect::eval_select(enquo(names_from), sim_data))
+  values_from <- names(tidyselect::eval_select(enquo(values_from), sim_data))
+
+  id_cols <- enquo(id_cols)
+  if (quo_is_null(id_cols)) {
+    sim_vars <- names(sim_data)
+    id_cols <- sim_vars[!sim_vars %in% c(names_from, values_from)]
+  } else {
+    id_cols <- names(tidyselect::eval_select(id_cols, sim_data))
+  }
+
+  if (length(names_from) > 1) {
+    new_vars <- mutate(shallow_dt(data), .names_from = paste(!!!syms(names_from), sep = names_sep))
+    new_vars <- unique(pull(new_vars, .names_from))
+  } else {
+    new_vars <- unique(pull(data, !!sym(names_from)))
+  }
+
+  if (!is.null(names_glue)) {
+    glue_df <- as.data.table(distinct(ungroup(data), !!!syms(names_from)))
+    glue_df <- vctrs::vec_rep(glue_df, length(values_from))
+    glue_df$.value <- vctrs::vec_rep_each(values_from, length(new_vars))
+    glue_vars <- as.character(glue::glue_data(glue_df, names_glue))
+  }
+
+  if (length(values_from) > 1) {
+    new_vars <- lapply(values_from, function(.x) paste(.x, new_vars, sep = names_sep))
+    new_vars <- unlist(new_vars)
+  }
+
+  no_id <- length(id_cols) == 0
+
+  if (no_id) {
+    lhs <- "..."
+    new_vars <- c(".", new_vars)
+  } else {
+    lhs <- paste(id_cols, collapse = " + ")
```

Comment:
Is there a reason to make these strings rather than calls? The main downside of using strings is that there's a high likelihood of weirdness in the presence of non-syntactic names.

Reply:
I think the change I made here covers what you were looking for, let me know if I misunderstood.

In general all changes are implemented and ready for review.